Sources/PackageGraph/CompositePackageContainerProvider.swift

```diff
@@ -0,0 +1,84 @@
+    /// the container.
+    let currentToolsVersion: ToolsVersion
+
+    /// The tools version loader.
+    let toolsVersionLoader: ToolsVersionLoaderProtocol
+
+    let fileSystem: FileSystem
+
+    /// Create a repository-based package provider.
+    ///
+    /// - Parameters:
+    ///   - repositoryManager: The repository manager responsible for providing repositories.
+    ///   - manifestLoader: The manifest loader instance.
+    ///   - currentToolsVersion: The current tools version in use.
+    ///   - toolsVersionLoader: The tools version loader.
+    public init(
+        repositoryManager: RepositoryManager,
+        mirrors: DependencyMirrors = DependencyMirrors(),
+        manifestLoader: ManifestLoaderProtocol,
+        currentToolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
+        toolsVersionLoader: ToolsVersionLoaderProtocol = ToolsVersionLoader(),
+        fileSystem: FileSystem = localFileSystem
+    ) {
+        self.repositoryManager = repositoryManager
+        self.mirrors = mirrors
+        self.manifestLoader = manifestLoader
+        self.currentToolsVersion = currentToolsVersion
+        self.toolsVersionLoader = toolsVersionLoader
+        self.fileSystem = fileSystem
+    }
+
+    public func getContainer(
+        for reference: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, Swift.Error>) -> Void
+    ) {
+        let result: Result<PackageContainer, Swift.Error>
+
+        // For remote package references, attempt to load from registry before falling back to respository access.
+        if reference.kind == .remote {
+            do {
+                let provider = RegistryPackageContainerProvider(mirrors: mirrors, manifestLoader: manifestLoader, currentToolsVersion: currentToolsVersion, toolsVersionLoader: toolsVersionLoader)
+                let container = try tsc_await { provider.getContainer(for: reference, skipUpdate: skipUpdate, completion: $0) }
+                result = .success(container)
+            } catch {
+                let provider = RepositoryPackageContainerProvider(repositoryManager: self.repositoryManager, mirrors: self.mirrors, manifestLoader: self.manifestLoader, currentToolsVersion: self.currentToolsVersion, toolsVersionLoader: self.toolsVersionLoader)
+                result = tsc_await { provider.getContainer(for: reference, skipUpdate: skipUpdate, completion: $0) }
+            }
+        } else {
+            let provider = LocalPackageContainerProvider(mirrors: self.mirrors, manifestLoader: self.manifestLoader, currentToolsVersion: self.currentToolsVersion, toolsVersionLoader: self.toolsVersionLoader, fileSystem: self.fileSystem)
+            result = tsc_await { provider.getContainer(for: reference, skipUpdate: skipUpdate, completion: $0) }
+        }
+
+        callbacksQueue.async {
+            completion(result)
+        }
+    }
```

Comment:
these ^^ should not "await", instead they should pass-through the completion handler or handler the failure case in the first case